[AArch64][GlobalISel] Legalize Insert vector element

davemgreen opened the pull request:
This attempts to standardize and extend some of the insert vector element lowering. Most notably:
 - More types are handled by splitting illegal vectors
 - The index type for G_INSERT_VECTOR_ELT is canonicalized to TLI.getVectorIdxTy(), similar to extact_vector_element.
 - Some of the existing patterns now have the index type specified to make sure they can apply to GISel too.
 - The C++ selection code has been removed, relying on tablegen patterns.
 - G_INSERT_VECTOR_ELT with small GPR input elements are pre-selected to use a i32 type, allowing the existing patterns to apply.
 - Variable index inserts are lowered in post-legalizer lowering, expanding into a stack store and reload.

arsenm commented on llvm/include/llvm/Target/GlobalISel/SelectionDAGCompat.td:
```diff
@@ -141,6 +141,7 @@ def : GINodeEquiv<G_CTLZ_ZERO_UNDEF, ctlz_zero_undef>;
 def : GINodeEquiv<G_CTTZ_ZERO_UNDEF, cttz_zero_undef>;
 def : GINodeEquiv<G_CTPOP, ctpop>;
 def : GINodeEquiv<G_EXTRACT_VECTOR_ELT, extractelt>;
+def : GINodeEquiv<G_INSERT_VECTOR_ELT, vector_insert>;
```
vector_insert is the deprecated form with lax type constraints. insertelt should be preferred

davemgreen replied:
I'm not sure that's really true though, from what I can tell. I know there are comments saying it is deprecated, but an i8 isn't a legal type in AArch64 - we want it to be an i32 inserted into a v16i8 vector. It's the same as buildvectors, where the operands can be truncated to allow legal types to be inserted into a vector with smaller elements. Multiple other backends still use vector_insert.

We could probably be using insertelt a little more in the aarch64 backend where we know the types should match, so maybe it would be good to have both as Equivs, but we don't want to be using it everywhere.

aemerson commented on llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:
```diff
@@ -2121,6 +2120,31 @@ bool AArch64InstructionSelector::preISelLower(MachineInstr &I) {
     return false;
   }
+  case TargetOpcode::G_INSERT_VECTOR_ELT: {
+    Register InsReg = I.getOperand(2).getReg();
```
Please add a comment documenting what this is trying to do and why.

davemgreen replied:
I've added a comment and rebased - thanks.

davemgreen commented:
Ping - I've rebased and fixed the RISCV tests. Thanks.

arsenm commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -2972,7 +2972,22 @@ bool IRTranslator::translateInsertElement(const User &U,
   Register Val = getOrCreateVReg(*U.getOperand(0));
   Register Elt = getOrCreateVReg(*U.getOperand(1));
-  Register Idx = getOrCreateVReg(*U.getOperand(2));
+  const auto &TLI = *MF->getSubtarget().getTargetLowering();
```
the TargetLowering should probably be an IRTranslator member if it isn't already

arsenm commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -2972,7 +2972,22 @@ bool IRTranslator::translateInsertElement(const User &U,
+  Register Idx;
+  if (auto *CI = dyn_cast<ConstantInt>(U.getOperand(2))) {
+    if (CI->getBitWidth() != PreferredVecIdxWidth) {
+      APInt NewIdx = CI->getValue().zextOrTrunc(PreferredVecIdxWidth);
+      auto *NewIdxCI = ConstantInt::get(CI->getContext(), NewIdx);
```
I would expect the IRTranslator to pass through the original IR type, and let the legalizer do this casting

davemgreen replied:
This is the same as translateExtractElement below. I think it's simpler if all index operands for insert/extract are of size getVectorIdxTy(), no matter where they get created.

arsenm replied:
Then that's also wrong. This is doing legalization work in the IRTranslator

arsenm replied:
The index type in the DAG is not always getVectorIdxTy. The DAG builder passes through the type as-is

davemgreen replied:
SelectionDAGBuilder will extend the types as the nodes get constructed in https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L3963.
It can be seen in https://godbolt.org/z/eWe1hvjG5.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:
```diff
@@ -2121,6 +2120,36 @@ bool AArch64InstructionSelector::preISelLower(MachineInstr &I) {
     return false;
   }
+  case TargetOpcode::G_INSERT_VECTOR_ELT: {
+    // The tablegen patterns for integer insert_vector_elts with small elements
+    // will expect the operand to be legalized to a i32 type. This extends the
+    // types of i8 and i16 element inserts, so that the existing tablegen
+    // patterns can apply.
```
This is doing legalization work. The selector code should be as simple as possible (i.e., you shouldn't have to preprocess anything for the patterns to work)

davemgreen replied:
I think this needs to be after reg-bank selection. Otherwise there isn't a good way to detect INSERT_VECTOR_ELT from gpr values vs fpr registers for the same type size.
The process is irtranslation -> opt -> legalization -> opt -> regbank select -> regbank-dependant legalization -> selection. And currently that lives in this preISelLower method. I'm not sure if it's the best place for it, but there are other operations like select that could do with post-reg-back-select legalization to treat grp selects differently to fpr.

arsenm replied:
regbank-dependant legalization is part of regbankselect; the apply mapping should apply any regbank related legalization work

davemgreen replied:
If it can avoid this manual selection of the ANYEXT then it sounds good. There didn't seem to be anything already like it in AArch64, but I've added something for it. Thanks.

madhur13490 commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -475,6 +476,59 @@ void applyEXT(MachineInstr &MI, ShuffleVectorPseudo &MatchInfo) {
+      Builder.buildPtrAdd(MRI.getType(StackTemp.getReg(0)), StackTemp, Mul)
+          .getReg(0);
+
+  if (InsertVal) {
```
InsertVal is set to true. Why can't this be unconditional then? Is this function missing something?

madhur13490 commented on llvm/test/CodeGen/AArch64/insertextract.ll:
```diff
@@ -111,25 +85,51 @@ entry:
+; CHECK-GI-NEXT:    mov d1, v0.d[1]
+; CHECK-GI-NEXT:    // kill: def $d0 killed $d0 killed $q0
+; CHECK-GI-NEXT:    mov sp, x29
+; CHECK-GI-NEXT:    ldp x29, x30, [sp], #16 // 16-byte Folded Reload
```
Do we know why there is codegen difference between SDAG and Global ISel?

madhur13490 commented on llvm/lib/Target/AArch64/AArch64InstrFormats.td:
```diff
@@ -7997,8 +7997,8 @@ class SIMDInsFromElement<string size, ValueType vectype,
-                 (elttype (vector_extract (vectype V128:$Rn), idxtype:$idx2)),
-                 idxtype:$idx))]>;
+                 (elttype (vector_extract (vectype V128:$Rn), (i64 idxtype:$idx2))),
+                 (i64 idxtype:$idx)))]>;
```
May I know why these updates are needed?

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:
```diff
@@ -407,6 +411,14 @@ void AArch64RegisterBankInfo::applyMappingImpl(
+  case TargetOpcode::G_INSERT_VECTOR_ELT: {
+    // Extend smaller gpr operands to 32 bit.
+    Builder.setInsertPt(*MI.getParent(), MI.getIterator());
+    auto Ext = Builder.buildAnyExt(LLT::scalar(32), MI.getOperand(2).getReg());
```
Could you move this code into the legalizer? We only state legal types for index 0. There are no references to 1 and 2.

davemgreen replied:
This can't be part of the legalizer for the reasons already mentioned already - the types we want depend on which register bank was selected for the inserted element.

tschuett replied:
MIR after the legalizer must be legal. There cannot be i8 or i16 types. They are plain illegal. We have to fix the legalizer to prevent illegal types.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -883,9 +883,15 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
-      .legalIf(typeInSet(0, {v16s8, v8s8, v8s16, v4s16, v4s32, v2s32, v2s64}))
+      .legalIf(
+          typeInSet(0, {v16s8, v8s8, v8s16, v4s16, v4s32, v2s32, v2s64, v2p0}))
       .moreElementsToNextPow2(0)
```
`.legalFor({{v16s8, s32, s64}, {v8s8, s32, s64}, {v8s16, s32, s64}, {v4s16, s32, s64}, {v4s32, s32, s64}, {v2s32, s32, s64}, {v2s64, s64, s64}, {v2p0, p0, s64}})`

davemgreen replied:
Those would not really be correct for fp types.

tschuett replied:
```
// Generic insertelement.
def G_INSERT_VECTOR_ELT : GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins type0:$src, type1:$elt, type2:$idx);
  let hasSideEffects = false;
}
```
We have to specify all three types in the legalizer: in/out vector, the element, and the index.

tschuett replied:
Ah, the famous f16 again.

tschuett replied:
We need floats in the LLT. `{v8i16, i16, i64}` is illegal. There are no i16s. `{v8f16, f16, i64}` is illegal because of the famous f16.

tschuett replied:
There is no notion of fpr and gpr in the legalizer. We have to state the legal types for operations. The issue comes up with untyped operations, like e.g., G_SELECT, G_INSERT_VETOR_ELT, G_BUILD_VECTOR, ... 

For typed operations, we need no floats in the LLT. s16 is legal for G_FADD, but it is not legal for G_ADD. In the later case, it would have been an i16.

davemgreen replied:
> There is no notion of fpr and gpr in the legalizer. We have to state the legal types for operations. The issue comes up with untyped operations, like e.g., G_SELECT, G_INSERT_VETOR_ELT, G_BUILD_VECTOR, ...
> 
> For typed operations, we need no floats in the LLT. s16 is legal for G_FADD, but it is not legal for G_ADD. In the later case, it would have been an i16.

I'm not sure I am understanding your comment, but I think I agree with the first half.

For the second half the problem is not between G_FADD and G_ADD, it is between G_FADD with fp16 and G_FADD with bf16. They both have the same s16 sized operands, but there is currently no way to tell them apart in GISel.

davemgreen commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -475,6 +476,54 @@ void applyEXT(MachineInstr &MI, ShuffleVectorPseudo &MatchInfo) {
+  Register Offset = MI.getOperand(3).getReg();
+  LLT VecTy = MRI.getType(MI.getOperand(0).getReg());
+  LLT EltTy = MRI.getType(MI.getOperand(2).getReg());
+  LLT IdxTy = MRI.getType(MI.getOperand(3).getReg());
```
I was hoping to make this method useful for both G_INSERT_VECTOR_ELT and G_EXTRACT_VECTOR_ELT in a follow-up, but have made it use the GInsertVectorElement in this one.

davemgreen committed:
[AArch64][GlobalISel] Legalize Insert vector element

This attempts to standardize and extend some of the insert vector element
lowering. Most notably:
 - More types are handled by splitting illegal vectors
 - The index type for G_INSERT_VECTOR_ELT is canonicalized to
   TLI.getVectorIdxTy(), similar to extact_vector_element.
 - Some of the existing patterns now have the index type specified to make sure
   they can apply to GISel too.
 - The C++ selection code has been removed, relying on tablegen patterns.
 - G_INSERT_VECTOR_ELT with small GPR input elements are pre-selected in
   reg-bank-select to use a i32 type, allowing the existing patterns to apply.
 - Variable index inserts are lowered in post-legalizer lowering, expanding
   into a stack store and reload.
